jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java

```diff
@@ -72,79 +153,165 @@ public void preConfigure(final WebAppContext context) throws Exception
+     * 
+     * To find them, examine the classloaders in the hierarchy above the
+     * webapp classloader that are URLClassLoaders. For jdk-9 we also
+     * look at the java.class.path, and the jdk.module.path.
+     * 
+     * @param context the WebAppContext being deployed
+     * @throws Exception
+     */
+    public void findAndFilterContainerPaths (final WebAppContext context)
+    throws Exception
+    {
+        //assume the target jvm is the same as that running
+        int targetPlatform = JavaVersion.VERSION.getPlatform();
+        //allow user to specify target jvm different to current runtime
+        Object target = context.getAttribute(JavaVersion.JAVA_TARGET_PLATFORM);
+        if (target!=null)
+            targetPlatform = Integer.valueOf(target.toString()).intValue();
+        
+        //Apply an initial name filter to the jars to select which will be eventually
+        //scanned for META-INF info and annotations. The filter is based on inclusion patterns.
+        String tmp = (String)context.getAttribute(CONTAINER_JAR_PATTERN);
+        Pattern containerPattern = (tmp==null?null:Pattern.compile(tmp));
+        ContainerPathNameMatcher containerPathNameMatcher = new ContainerPathNameMatcher(context, containerPattern);
+        
         ClassLoader loader = null;
         if (context.getClassLoader() != null)
             loader = context.getClassLoader().getParent();
 
+        List<URI> containerUris = new ArrayList<>();
+        
         while (loader != null && (loader instanceof URLClassLoader))
         {
             URL[] urls = ((URLClassLoader)loader).getURLs();
             if (urls != null)
             {
-                URI[] containerUris = new URI[urls.length];
-                int i=0;
                 for (URL u : urls)
                 {
                     try
                     {
-                        containerUris[i] = u.toURI();
+                        containerUris.add(u.toURI());
                     }
                     catch (URISyntaxException e)
                     {
-                        containerUris[i] = new URI(u.toString().replaceAll(" ", "%20"));
+                        containerUris.add(new URI(u.toString().replaceAll(" ", "%20")));
                     }
-                    i++;
                 }
-                containerJarNameMatcher.match(containerPattern, containerUris, false);
             }
             loader = loader.getParent();
         }
+        
+        if (LOG.isDebugEnabled()) LOG.debug("Matching container urls {}", containerUris);
+        containerPathNameMatcher.match(containerUris);
 
-        //Apply ordering to WEB-INF/lib jars
-        PatternMatcher webInfJarNameMatcher = new PatternMatcher ()
+        //if running on jvm 9 or above, we we won't be able to look at the application classloader
+        //to extract urls, so we need to examine the classpath instead.
+        if (JavaVersion.VERSION.getMajor() >= 9)
```

Comment:
Major or Platform?

Reply:
I picked JavaVersion.getMajor() because I don't care about the exact version number (ie 1.6,1.7,1.8), I just care whether it is 9 or earlier (getMajor() returns 1 if < jdk9).

Reply:
But for consistency I'll change to getPlatform()